File: typescript/gql-codegen/src/schema/client/fragment.ts
/* eslint-disable max-lines */
import { z } from 'zod/v4'
import { argument } from './argument.js'

type UnionFragmentSpec = z.ZodObject<{
    _type: z.ZodLiteral<'UnionFragmentSpec'>;
    name: z.ZodString;
    selections: z.ZodArray<typeof unionSelection>
}>
type ObjectFragmentSpec = z.ZodObject<{
    _type: z.ZodLiteral<'ObjectFragmentSpec'>;
    name: z.ZodString;
    selections: z.ZodArray<typeof objectSelection>
}>
export type FragmentSpecSchemaZodType = z.ZodDiscriminatedUnion<[
    ObjectFragmentSpec,
    UnionFragmentSpec
]>
export type FragmentSpecSchemaType =
    {
        _type: 'UnionFragmentSpec',
        name: string,
        selections: UnionSelection[]
    } |
    {
        _type: 'ObjectFragmentSpec',
        name: string,
        selections: z.infer<typeof objectSelection>[]
    }

export const fieldSelection =
    z.object({
        _type: z.literal('FieldSelection'),
        name: z.string(),
        alias: z.string(),
        arguments: z.record(z.string(), argument),
        get selection(): z.ZodNullable<FragmentSpecSchemaZodType> {
            // eslint-disable-next-line no-use-before-define
            return z.nullable(fragmentSpecSchema)
        }
    })

export const typenameSelection = z.object({
    _type: z.literal('TypenameField'),
    alias: z.nullable(z.string())
})

export const objectConditionalSpreadSelection = z.object({
    _type: z.literal('ObjectConditionalSpreadSelection'),
    object: z.string(),
    get spec(): ObjectFragmentSpec {
        // eslint-disable-next-line no-use-before-define
        return objectFragmentSpec
    }
})

export const spreadSelection = z.object({
    _type: z.literal('SpreadSelection'),
    fragment: z.string()
})

export type UnionSelection =
    z.infer<typeof typenameSelection> |
    z.infer<typeof spreadSelection> |
    z.infer<typeof objectConditionalSpreadSelection> |
    z.infer<typeof unionConditionalSpreadSelection>

export type UnionSelectionZodType = z.ZodLazy<
    z.ZodDiscriminatedUnion<[
        typeof typenameSelection,
        typeof spreadSelection,
        typeof objectConditionalSpreadSelection,
        typeof unionConditionalSpreadSelection
    ]>
>


export const unionConditionalSpreadSelection = z.object({
    _type: z.literal('UnionConditionalSpreadSelection'),
    union: z.string(),
    get selections(): z.ZodArray<UnionSelectionZodType> {
        // eslint-disable-next-line no-use-before-define
        return z.array(unionSelection)
    }
})

export const unionSelection: UnionSelectionZodType =
    z.lazy(() => z.discriminatedUnion('_type', [
        typenameSelection,
        spreadSelection,
        objectConditionalSpreadSelection,
        unionConditionalSpreadSelection,
    ]))

export const objectSelection = z.lazy(() => z.discriminatedUnion('_type', [
    fieldSelection,
    typenameSelection,
    spreadSelection
]))

export const objectFragmentSpec: ObjectFragmentSpec =
    z.object({
        _type: z.literal('ObjectFragmentSpec'),
        name: z.string(),
        selections: z.array(objectSelection)
    })

export const unionFragmentSpec =
    z.object({
        _type: z.literal('UnionFragmentSpec'),
        name: z.string(),
        selections: z.array(unionSelection)
    })

export const fragmentSpecSchema =
    z.discriminatedUnion('_type', [
        objectFragmentSpec,
        unionFragmentSpec
    ]) satisfies FragmentSpecSchemaZodType

export const fragmentSchema = z.object({
    sourceText: z.string(),
    spec: fragmentSpecSchema
})

File: rust/gqlcodegen/src/schema/client/fragment.rs
use std::collections::HashMap;

use crate::schema::client::argument::Argument;

#[derive(Debug, serde::Deserialize)]
pub struct FieldSelection {
    pub name: String,
    pub alias: String,
    pub arguments: HashMap<String, Argument>,
    pub selection: Option<FragmentSpec>
}

#[derive(Debug, serde::Deserialize)]
pub struct TypenameSelection {
    pub alias: Option<String>
}

#[derive(Debug, serde::Deserialize)]
pub struct SpreadSelection {
    pub fragment: String
}

#[derive(Debug, serde::Deserialize)]
#[serde(tag = "_type")]
pub enum ObjectSelection {
    #[serde(rename(deserialize = "FieldSelection"))]
    Field(FieldSelection),
    #[serde(rename(deserialize = "TypenameField"))]
    Typename(TypenameSelection),
    #[serde(rename(deserialize = "SpreadSelection"))]
    Spread(SpreadSelection)
}



#[derive(Debug, serde::Deserialize)]
#[serde(tag = "_type")]
pub enum UnionSelection {
    #[serde(rename(deserialize = "ObjectConditionalSpreadSelection"))]
    ObjectConditionalSpread {
        object: String,
        spec: ObjectFragmentSpec
    },
    #[serde(rename(deserialize = "UnionConditionalSpreadSelection"))]
    UnionConditionalSpread {
        union: String,
        selections: Vec<UnionSelection>
    },
    #[serde(rename(deserialize = "TypenameField"))]
    Typename(TypenameSelection),
    #[serde(rename(deserialize = "SpreadSelection"))]
    Spread(SpreadSelection)
}

#[derive(Debug, serde::Deserialize)]
#[serde(tag = "_type")]
pub struct ObjectFragmentSpec {
    pub name: String,
    pub selections: Vec<ObjectSelection>
}

#[derive(Debug, serde::Deserialize)]
#[serde(tag = "_type")]
pub enum FragmentSpec {
    #[serde(rename(deserialize = "ObjectFragmentSpec"))]
    Object(ObjectFragmentSpec),
    #[serde(rename(deserialize = "UnionFragmentSpec"))]
    Union{
        name: String,
        selections: Vec<UnionSelection>
    }
}

#[derive(Debug, serde::Deserialize)]
pub struct Fragment {
    #[serde(rename(deserialize = "sourceText"))]
    source_text: String,
    spec: FragmentSpec
}


